### PR TITLE
Patched the combine_results function, to reduce the double single_ep …

### PR DESCRIPTION
**Pull request by p0psicles**
…results.

We only need one (best) result using the sort_results function.
* Added __eq__ to episode, so the == checks work properly on episode objects.

- [x] PR is based on the DEVELOP branch
- [x] Don't send big changes all at once. Split up big PRs into multiple smaller PRs that are easier to manage and review
- [x] Read the [contribution guide](https://github.com/pymedusa/Medusa/blob/master/.github/CONTRIBUTING.md)

### CODE REVIEW

File: medusa/search/core.py [hunk 1]
```diff
@@ -906,6 +906,15 @@ def combine_results(multi_results, single_results):
     """Combine single and multi-episode results, filtering out overlapping results."""
     log.debug(u'Combining single and multi-episode results')
     result_candidates = []
+    return_single_results = []
+    return_multi_results = []
+
+    # Only return one result per episode for the single results and multi-result result_candidates
+    if single_results:
```
Comment (p0psicles) [inline medusa/search/core.py]: @medariox this is now added to go back from a list like: `[61, 52, 61, 11, 11, 61, 61]` to `[61, 52, 11]` (the numbers are now episode objects of course, which can be compared to one another.

File: medusa/search/core.py [hunk 2]
```diff
@@ -951,16 +960,18 @@ def combine_results(multi_results, single_results):
...
                     log.debug(
                         u'A needed multi-episode result overlaps with a single-episode result'
                         u' for episode {0}, removing the single-episode results from the list',
                         ep_obj.episode,
                     )
-                    del single_results[i]
+                    del return_single_results[i]
 
-    return single_results + result_candidates
+    return return_single_results + return_multi_results
```
Comment (p0psicles) [inline medusa/search/core.py]: If this method does its work. A result for a specific shows episode, should only be in `return_single_results` *or* in `return_multi_results` but never in both. Which was the case now.

File: medusa/search/core.py [hunk 3]
```diff
@@ -938,8 +947,8 @@ def combine_results(multi_results, single_results):
         # see how many of the eps that this result covers aren't covered by single results
         needed_eps = []
         not_needed_eps = []
-        for result in single_results:
-            if result in multi_result.episodes:
+        for result in return_single_results:
+            if result.episodes[0] in multi_result.episodes:
```
Comment (p0psicles) [inline medusa/search/core.py]: Previously you where comparing are TorrentSearch object with a list of episode objects. Which was always `False` of course.

File: medusa/search/core.py [hunk 4]
```diff
@@ -792,6 +778,7 @@ def collect_candidates(found_results, provider, multi_results, single_results, s
                   if result in (SEASON_RESULT, MULTI_EP_RESULT))
     candidates = list(itertools.chain(*candidates))
     if candidates:
+        # FIXME: collect_multi_candidates can potentially result in downloading multiple multi-ep results.
```
Comment (medariox) [inline medusa/search/core.py]: This needs to be handled in collect_results as we need all single episode results and all multi episode results to determine the needed episode results.

Comment (medariox) [reply medusa/search/core.py]: And it should be already working like that.

Comment (medariox) [reply medusa/search/core.py]: Yes, because the filtering is done in combine_results as we need the single episode results.

Comment (p0psicles) [reply medusa/search/core.py]: I'm saying that for example, nyaa returns one release that's multi_ep. But after this method, it's duplicated? Why?

**Pull request merged**